File: src/main/java/core/procedures/functional/ForEach.java
package core.procedures.functional;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.procedures.IFn;
import core.scm.Cons;
import core.scm.Type;
import core.scm.Thunk;

public final class ForEach extends AFn {

  public ForEach() {
    super(new FnArgsBuilder().min(2).mandatory(new Class[]{IFn.class}).rest(Type.ProperList.class).build());
  }

  @Override
  public String getName() {
    return "for-each";
  }

  @Override
  public Thunk apply(Object... args) {
    /* For-each is the same as Map, but ignores the result */
    Thunk map = MapProc.MAP_PROC.apply(args);
    /* Void (ignore) results: (void <map-results>) */
    return new Thunk(Cons.list(VoidProc.VOID, map.getExpr()));
  }
}

File: src/main/java/core/procedures/functional/ForEach.kt
package core.procedures.functional

import core.procedures.AFn
import core.procedures.Arity.AtLeast
import core.procedures.IFn
import core.scm.Thunk

/* For-each is the same as `map`, but ignores the result: Void (ignore) results: (void <map-results>) */
class ForEach : AFn<Any?, Any?>(name = "for-each", arity = AtLeast(2), mandatoryArgsTypes = arrayOf(IFn::class.java)) {

    private val map = object : MapProc() { override val name = "for-each" }
    private val void = VoidProc()
    private val apply = Apply()

    override operator fun invoke(args: Array<out Any?>) = Thunk(listOf(apply, void, map(args)))
}